### PR TITLE
Restructure charms client

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Create new LocalCharmClient, which steals the method AddLocalCharm from
the original charms client.

This is because in future we will want to add extra dependencies to
AddLocalCharm, which are not relevant to the rest of the api methods.

(We are soon to add an S3-compatible endpoint to the controller for
uploading local charms, meaning we will need to inject an S3 client into
the api so that this Put operation can be performed)

Only a small number of the charm.Clients we construct actually need to
be able to call AddLocalCharm. Only two in production code (once in deploy,
once in refresh)

All non-deploy/refresh code (e.g. diffbundle, ssh_container,
charmresources) will just keep the reduced client, meaning they never
need to worry about these extra dependencies soon to be introduced
## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Since this is quite a simple change, simply compiling should do. However, for completeness:

```
$ juju bootstrap lxd lxd
$ juju add-model m
$ juju deploy ubuntu
$ juju refresh ubuntu --path /path/to/charm

$ juju add-model m2
$ juju deploy juju-qa-bundle-test
$ juju diff-bundle juju-qa-bundle-test

$ juju add-model m3
$ juju deploy ~/charms/ubuntu

$ juju add-model m4
$ cat ./bundle.yaml
applications:
  ubuntu:
    charm: /gome/jack/charms/ubuntu
$ juju deploy ./bundle.yaml
```

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
Much better, I'll run the QA steps now.

**Review by hmlanigan — CHANGES_REQUESTED** [review]
After chats, we're looking to have an `charms.AddCharmAdapter(charms.NewClient(apiroot), Uploader)`, where uploader can be for s3, http, or nil if `AddLocalCharm` is not called.  The adapter would cover both AddLocalCharm and AddCharm. This looks okay in the juju client code and for the terraform provider for juju, which doesn't support local charm deployments.

@SimonRichardson this is a small change from our prior discussions.

The PR and commit messages will have enhanced descriptions to reduce ambiguity.

**Comment by SimonRichardson** [comment]
@hmlanigan we should never pass nil. An error should suffice if nil uploader. If we allow nil, then I see no point in this pr. We should just push the uploader in to the existing client.

**Comment by SimonRichardson** [comment]
Also that's why I wanted `AddCharm` to be on the existing client. Uploading local charms is the anomaly here, so you should opt into it and the new adapter pattern gives us that. The terraform provider will not use the adapter, as it doesn't support local charms.

**Comment by jack-w-shaw** [comment]
Following discussions, we have decided instead to break out AddLocalCharm into it's own LocalCharm client

File: api/client/charms/local_charm_client.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	base.ClientFacade
+	facade base.FacadeCaller
+}
+
+func NewLocalCharmClient(st base.APICallCloser) *LocalCharmClient {
+	frontend, backend := base.NewClientFacade(st, "Charms")
+	return &LocalCharmClient{ClientFacade: frontend, facade: backend}
+}
+
+// AddAddCharmAdaptor prepares the given charm with a local: schema in its
```
Comment (hmlanigan) [inline api/client/charms/local_charm_client.go:37]: s/AddAddCharmAdaptor/AddLocalCharm/

File: api/client/charms/local_charm_client.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
+// Copyright 2023 Canonical Ltd.
```
Comment (hmlanigan) [inline api/client/charms/local_charm_client.go:1]: The filenames should be `localcharmclient.go` and `localcharmclient_test.go`

**Review by hmlanigan — APPROVED** [review]
2 small changes and it's approved.

**Pull request merged**